File: c++/LongestIncreasingSubsequence.cpp
class Solution {
public:
    int lengthOfLIS(vector<int>& nums) {
        
        vector<int> sub;
        
        int n = nums.size();
        
        for(int i = 0; i < n; i++) {
            auto it = lower_bound(sub.begin(), sub.end(), nums[i]); //returns index of value found that 
compares less than the value nums[i]
            if(it != sub.end()) {
                *it = nums[i]; //We change the value found
            } else {
                sub.push_back(nums[i]); //We can now safely move our iterator up after pushing back
            }
        }
        for(auto a : sub)
            cout << a << endl;
        return sub.size();
    }
};

File: java/LongestIncreasingSubSequence.java
class Solution {

    public int lengthOfLIS(int[] nums) {
       
        int n = nums.length;
         int dp []  = new int [n+1];
         for(int i =0 ; i < dp.length;i++)
            dp[i] = 1;
        for(int i = n-1; i >= 0;i--)
        {
            for(int j = i; j < n;j++){
                if(nums[i] < nums[j])
                    dp[i] = Math.max(dp[i],dp[j] + 1);
            }
        }
        int maxSeq = 0;
        for(int i =0; i < n;i++)
            maxSeq = Math.max(maxSeq,dp[i]);
        for(int i =0;i < n;i++)
            System.out.print(dp[i] + ",");
        return maxSeq;
    }
}